ctrd/container.go

```diff
@@ -137,50 +132,68 @@ func (c *Client) execContainer(ctx context.Context, process *Process) error {
 		return errors.Wrap(err, "failed to exec process")
 	}
 
-	errCh := make(chan error, 1)
-	defer close(errCh)
-
-	go func() {
-		var msg *Message
+	var msg *Message
+	defer func() {
+		if msg != nil {
+			// XXX: if exec process get run, io should be closed in this function,
+			for _, hook := range c.hooks {
+				if err := hook(process.ExecID, msg); err != nil {
+					logrus.Errorf("failed to execute the exec exit hooks: %v", err)
+					break
+				}
+			}
 
-		if startErr := <-errCh; startErr != nil {
-			msg = &Message{
-				err:      startErr,
-				exitCode: 126,
-				exitTime: time.Now().UTC(),
+			// delete the finished exec process in containerd
+			if _, err := execProcess.Delete(context.TODO()); err != nil {
+				logrus.Warnf("failed to delete exec process %s: %s", process.ExecID, err)
 			}
 		}
+	}()
 
-		// success to start which means the cmd is valid and wait
-		// for process.
-		if msg == nil {
-			status := <-exitStatus
-			msg = &Message{
-				err:      status.Error(),
-				exitCode: status.ExitCode(),
-				exitTime: status.ExitTime(),
-			}
+	// start the exec process
+	if err := execProcess.Start(ctx); err != nil {
+		msg = &Message{
+			err:      err,
+			exitCode: 126,
+			exitTime: time.Now().UTC(),
 		}
+		return errors.Wrapf(err, "failed to start exec, exec id %s", execID)
+	}
+	// make sure the closeStdinCh has been closed.
+	close(closeStdinCh)
+
+	t := time.Duration(timeout) * time.Second
+	var timeCh <-chan time.Time
+	if t == 0 {
+		timeCh = make(chan time.Time)
+	} else {
+		timeCh = time.After(t)
+	}
 
-		// XXX: if exec process get run, io should be closed in this function,
-		for _, hook := range c.hooks {
-			if err := hook(process.ExecID, msg); err != nil {
-				logrus.Errorf("failed to execute the exec exit hooks: %v", err)
-				break
+	select {
+	case status := <-exitStatus:
+		msg = &Message{
+			err:      status.Error(),
+			exitCode: status.ExitCode(),
+			exitTime: status.ExitTime(),
+		}
+	case <-timeCh:
+		// Ignore the not found error because the process may exit itself before kill
+		if err := execProcess.Kill(ctx, syscall.SIGKILL); err != nil && !errdefs.IsNotFound(err) {
+			//try to force kill the exec process
+			if err := execProcess.Kill(ctx, syscall.SIGTERM); err != nil && !errdefs.IsNotFound(err) {
+				logrus.Errorf("failed to kill the exec process, %v", err)
```

Comment:
we should return error here instead of warning?  cc @fuweid

Reply:
+1